relations/src/gr1cs/constraint_system.rs

```diff
@@ -0,0 +1,660 @@
+
+    /// Returns the number of instance variables.
+    pub fn num_instance_variables(&self) -> usize {
+        self.num_instance_variables
+    }
+
+    /// Returns the number of witness variables.
+    pub fn num_witness_variables(&self) -> usize {
+        self.num_witness_variables
+    }
+
+    /// Returns the number of local predicates
+    pub fn num_local_predicates(&self) -> usize {
+        self.local_predicates.len()
+    }
+
+    /// Enforce a constraint in the constraint system. It takes a local
+    /// predicate name and enforces a vector of linear combinations of the
+    /// length of the arity of the local predicate enforces the constraint.
+    #[inline]
+    pub fn enforce_constraint(
+        &mut self,
+        local_predicate_label: &str,
+        lc_vec: Vec<LinearCombination<F>>,
+    ) -> crate::gr1cs::Result<()> {
+        if !self.local_predicates.contains_key(local_predicate_label) {
+            return Err(SynthesisError::PredicateNotFound);
+        }
+        if self.should_construct_matrices() {
+            let constraint = self.new_lc_vec(lc_vec)?;
+            self.local_predicates
+                .get_mut(local_predicate_label)
+                .unwrap()
+                .enforce_constraint(constraint)?;
+        }
+        #[cfg(feature = "std")]
+        {
+            let trace = ConstraintTrace::capture();
+            match self.predicate_traces.get_mut(local_predicate_label) {
+                Some(traces) => traces.push(trace),
+                None => (),
+            }
+        }
+        Ok(())
+    }
+
+    /// Enforce an r1cs constraint in the constraint system. It takes a, b, and
+    /// c and enforces `a * b = c`. If R1CS predicate does not exist in the
+    /// constraint system, It will create one. This function is a special case
+    /// of `enforce_constraint` and is used as the legacy R1CS API to be bacward
+    /// compatible with R1CS gadgets
+    #[inline]
+    pub(super) fn enforce_r1cs_constraint(
+        &mut self,
+        a: LinearCombination<F>,
+        b: LinearCombination<F>,
+        c: LinearCombination<F>,
+        cs: ConstraintSystemRef<F>,
+    ) -> crate::gr1cs::Result<()> {
+        if !self.local_predicates.contains_key(R1CS_PREDICATE_LABEL) {
+            self.register_predicate(
+                R1CS_PREDICATE_LABEL,
+                LocalPredicate::build_r1cs_predicate(cs),
+            )?;
+        }
+        self.enforce_constraint(R1CS_PREDICATE_LABEL, vec![a, b, c])
+        // if self.should_construct_matrices() {
+        //     let constraints = self.new_lc_vec(vec![a, b, c])?;
+        //     self.local_predicates
+        //         .get_mut(R1CS_PREDICATE_LABEL)
+        //         .unwrap()
+        //         .enforce_constraint(constraints)?;
+        // }
+        // Ok(())
+    }
+
+    /// Add a new vector of linear combinations to the constraint system.
+    pub fn new_lc_vec(
+        &mut self,
+        lc_vec: Vec<LinearCombination<F>>,
```

Comment:
```suggestion
        lc_vec: impl IntoIterator<Item = LinearCombination<F>>,
```

I can pass in `[lc_1, lc_2, lc_3]`, or I can pass in `vec![lc_1, lc_2, lc_3]`. The benefit is that this avoids allocations